### ISSUE TITLE
Why and How does Rserve make "conn" folders on Linux? Is there a way to get it to clean up its directories?

### ISSUE THREAD

**Issue by mlane3** [s-u/Rserve#187]
Dear @s-u I have a hilarious bug that I need a bit of an advice on how to handle.   The picture shows the bug prehaps best:
[https://i.imgur.com/MVpuFGq.png](https://i.imgur.com/MVpuFGq.png)

- Question 1: R keeps creating these ""conn"" folders.   Do you know why it is making them? 
- Question 2: We are using RServe in combination with tidymodels to do MLOps and set up using chrontab.  So inherently we have programming situation where two packages have to share the same working environment.  I think I could write a shell script to clean things up but then I notice that in [rserv.conf notes](https://github.com/s-u/Rserve/wiki/rserve.conf) there are a bunch of new settings like *source*, *workdir.mode*, *workdir.parent.mode* -- all which have very interesting properties.  Do you have any suggestions on how my team could organize things so that RServe deletes the "conn" folders but keeps everything not in a conn folder?  I like to try these new features. 

Both rserv.cfg and /etc/rconf contain have these same settings: 
```
workdir /home/radmin/studentdir/WorkDir/
pwdfile /home/radmin/studentdir/WorkDir/pass.txt
remote enable
plaintext enable
interactive yes
port 6311
````
The script we use to start Rserve is: 
`R CMD Rserve --RS-enable-remote --no-save --RS-conf /home/radmin/studentdir/WorkDir/rserv.cfg > Rserve-$(date +"%Y_%m_%d_%I_%M>`

P.S. We are using RServe in combination with tidymodels to do ModelOps and MLOps.  I need Rserve and tidymodels to share the same root working directory so that when each is called using shell scripts they work appropriately.  I think I could write a shell script to clean things up, but I notice that there is a new function in RSClient package called

**Comment by s-u** [comment]
@mlane3 I'm not sure I understand your question, but `Rserve` creates distinct temporary directories for each connection to make sure they don't trample on each other when working, i.e. each connection is guaranteed to have a unique working directory. You can set the location using the `workdir` configuration directive (or `--RS-workdir` argument).

By default each working directory is automatically removed, but only is if it empty. The assumption is that the script should clean up what's not needed and if anything was left over, it was intended to be persistent so Rserve should not delete it. If you want to delete it regardless you can set the `workdir.clean` configuration. I would suggest looking at all the `workdir*` [configuration options](https://github.com/s-u/Rserve/wiki/rserve.conf) in case it's useful to you.

**Comment by mlane3** [comment]
So now I am confused myself.  Here are some more pictures of the log file latest when Rserve has been running and a second picture showing how all the connection folders are empty.  Do either of these image help make sense of what is going on? 

My confusion, because basically while running during the day our Tableau Server or Tableau Desktop makes a connection to Rserve.  But I have only tested the connection with one computer.  So I don't understand how I have 5 connection folders for 1 test.  Does Tableau make a new connection every single time?  I am starting to wonder if the "no-save" option is not working correctly?

![image](https://user-images.githubusercontent.com/1785885/189246648-0644171e-1fb7-4716-8984-6ae228c297e8.png)
![image](https://user-images.githubusercontent.com/1785885/189246692-3dd634b9-a2b1-4fd1-8702-d1e65e8e1e32.png)

**Comment by mlane3** [comment]
@s-u right so on those workspace options that were my real question lies. 
I am looking for if wordir.clean or workdir.parent.directory would give me a similar behavior to "--no-save"

For my questions please assume that the configuration is the same as above....
Here is what I found:
> added workdir.parent.mode option which specifies the
        permissions on the parent directory of session working
        directories. The default is 0755. Unlike workdir.mode this one
        does not respect the umask (beause it may be shared and thus
        may need to include permissions that are otherwise not
        granted).
and for work.clean
> added new option to remove the working directory on session close recursively - activated by  workdir.clean enable.  If not set, the default is to only remove it if it is empty (on the assumption there are files that will be served beyond the life of the session).

- My first question is I am a windows user so everything in linux is new to me.  What do you mean by "umask" when it comes to permissions? Why is it useful to change permissions for created folders? Could the permissions of a folder prevent them from being deleted? 
- My second question is with work.clean if "studentsdir/Workdir/ is configured as the working directory above... then if I use workdir.clean will it delete everything like the pass.txt from Workdir? Or does workdir.clean delete only everything from a conn folder like conn103626?

**Comment by s-u** [comment]
First, `--no-save` has absolutely nothing to do with `Rserve` - what R does is up to the code that the client evaluates. I cannot comment on what Tableau does - you'd have to ask the authors.

`workdir.clean yes` deletes the `conn...` directories unconditionally even if they are not empty. Your screenshots don't necessarily tell anything - if you want to see if there is something in a directory you'd have to use `ls -la` since there are likely hidden files.

**Comment by mlane3** [comment]
> First, `--no-save` has absolutely nothing to do with `Rserve` - what R does is up to the code that the client evaluates. I cannot comment on what Tableau does - you'd have to ask the authors.
Yeah, the original developers who created the connections for Tableau did not stay at the company long, so contacting them has been frustrating.  I eventually had to figure out things on my own (see 1).

> `workdir.clean yes` deletes the `conn...` directories unconditionally even if they are not empty. Your screenshots don't necessarily tell anything - if you want to see if there is something in a directory you'd have to use `ls -la` since there are likely hidden files.
Thanks for pointing that out.  A lot of them are still empty after I looked for hidden files.  There are new interns with Linux experience coming to the project, so I will make them add to the script a line to clean up any conn folders.   

(1) P.S.  Not sure if it's helpful, but did you know if you just copy the .dlls in the /R/ folder to the right folder in /Rserve/ then both your Rserve.exe and Rserved.exe files work just like they would in Linux?   It's basically a reverse of what you put in an FAQ, but it's what I have been using for all my organization's new hires.   Its also not scalable as those .dlls are both R version and machine specific, but for making things just a little easier for new R users it has been a godsend to give them their own mini-debug server
